Mathlib/FieldTheory/PurelyInseparable.lean

```diff
@@ -0,0 +1,644 @@
+    haveI := (isPurelyInseparable_adjoin_simple_iff_mem_pow F p (x := y)).2
+      ⟨1, x, by rw [← h, pow_one, frobenius_def]⟩
+    haveI := isSeparable_tower_bot_of_isSeparable F F⟮y⟯ E
+    obtain ⟨z, h'⟩ := F⟮y⟯.eq_bot_of_isPurelyInseparable_of_isSeparable.le <|
+      mem_adjoin_simple_self F y
+    exact ⟨z, (algebraMap F E).injective (by rw [← h, ← h', frobenius_def, map_pow]; rfl)⟩
+  exact PerfectRing.toPerfectField F p
+
+/-- If `E` is an algebraic closure of `F`, then `F` is perfect if and only if `E / F` is
+separable. -/
+theorem perfectField_iff_isSeparable_algebraicClosure [IsAlgClosure F E] :
+    PerfectField F ↔ IsSeparable F E :=
+  ⟨fun _ ↦ IsSepClosure.separable, fun _ ↦ haveI : IsAlgClosed E := IsAlgClosure.alg_closed F;
+    perfectField_of_isSeparable_of_perfectField_top F E⟩
+
+namespace Field
+
+/-- If `E / F` is algebraic, then the `Field.finSepDegree F E` is equal to `Field.sepDegree F E`
+as a natural number. This means that the cardinality of `Field.Emb F E` and the degree of
+`(separableClosure F E) / F` are both finite or infinite, and when they are finite, they
+coincide. -/
+theorem finSepDegree_eq (halg : Algebra.IsAlgebraic F E) :
+    finSepDegree F E = Cardinal.toNat (sepDegree F E) := by
+  have h := finSepDegree_mul_finSepDegree_of_isAlgebraic F (separableClosure F E) E
+    (halg.tower_top _) |>.symm
+  haveI := separableClosure.isSeparable F E
+  haveI := separableClosure.isPurelyInseparable F E halg
+  rwa [finSepDegree_eq_finrank_of_isSeparable F (separableClosure F E),
+    IsPurelyInseparable.finSepDegree_eq_one (separableClosure F E) E, mul_one] at h
+
+/-- The (finite) separable degree multiply by the (finite) inseparable degree is equal
+to the (finite) field extension degree. -/
+theorem finSepDegree_mul_finInsepDegree : finSepDegree F E * finInsepDegree F E = finrank F E := by
+  by_cases halg : Algebra.IsAlgebraic F E
+  · have := congr_arg Cardinal.toNat (sepDegree_mul_insepDegree F E)
+    rwa [Cardinal.toNat_mul, ← finSepDegree_eq F E halg] at this
+  change _ * finrank (separableClosure F E) E = _
+  rw [finrank_of_infinite_dimensional (K := F) (V := E) fun _ ↦
+      halg (Algebra.IsAlgebraic.of_finite F E),
+    finrank_of_infinite_dimensional (K := separableClosure F E) (V := E) fun _ ↦
+      halg ((separableClosure.isAlgebraic F E).trans (Algebra.IsAlgebraic.of_finite _ E)),
+    mul_zero]
+
+end Field
+
+namespace separableClosure
+
+variable [Algebra E K] [IsScalarTower F E K]
+
+/-- If `K / E / F` is a field extension tower, such that `E / F` is purely inseparable and `K / E`
+is separable, then `E` adjoin `separableClosure F K` is equal to `K`. It is a special case of
+`separableClosure.adjoin_eq_of_isAlgebraic`, and is an intermediate result used to prove it. -/
+lemma adjoin_eq_of_isPurelyInseparable' [IsPurelyInseparable F E] [IsSeparable E K] :
+    adjoin E (separableClosure F K : Set K) = ⊤ := top_unique fun x _ ↦ by
```

Comment:
![1704703968805](https://github.com/leanprover-community/mathlib4/assets/3064145/4a49fc71-aebd-4138-af55-d2346944f73b)

p.227 of Lang's Algebra introduces the notion of a distinguished class of extensions, of which the classes of algebraic, separable, and purely inseparable extensions are all examples. Condition (1) is basically `trans` and `tower_top/bot` lemmas, and Condition (3) is `sup` lemmas (follows from (1) and (2)). Condition (2) says if K and L are intermediate fields of E/F and if K/F is in the class, then KL/L is also in the class. In Lean this can be generalized to say that if E/L/F is a tower and K is an intermediate field of E/F such that K/F is in the class, then `adjoin L K` over `L` is in the class. This type of lemmas seem to be missing from mathlib at this moment and it looks like they may be useful in places like this.

Reply:
Looks good. In algebraic geometry there are also such distinguished classes I think.

Reply:
We have something similar. See for example `AlgebraicGeometry/Morphisms/Basic.lean` or `CategoryTheory/MorphismProperty.lean`